### ISSUE TITLE
"The plugin repository could not be contacted."

### ISSUE THREAD

**Issue by fiwswe** [splitbrain/dokuwiki#3148]
This may be related to #1407 ?

### Setup:
• OpenBSD 6.7
• Apache 2.4.43
• PHP 7.4.6 (+ php-apache, php-gd, php-mysqli, php-snmp packages)
• Dokuwiki 2018-04-22c "Greebo"
• The server has a public IP and outgoing traffic is not blocked. (Incoming traffic is restricted by a firewall but ports 80/tcp and 443/tcp are open.)
• Access to the wiki directory is limited using HTTP Basic authentication in the Apache conf file for the VirtualHost like this:
```
<VirtualHost xx.xx.xx.xx:80>
        ServerName _SERVER_FQDN_
        # Note: dokuwiki is installed directly in this directory so access is:
        # http://_SERVER_FQDN_/doku.php?id=start
        DocumentRoot /home/_UNIX_USERNAME_/www-wiki
        …
        <Directory /home/_UNIX_USERNAME_/www-wiki>
                AllowOverride ALL
                AuthType Basic
                AuthName "Admins only"
                AuthUserFile /var/www/conf/htpasswd-_xxx_
                Require user _user1_ _user2_ _user3_
        </Directory>
        …
```

### Issue:
When I open the Extension Manager (on the "Installed Plugins" or any other tab) http://_SERVER_FQDN_/doku.php?id=sidebar&do=admin&page=extension&tab=plugins page, I see the message:
> "The plugin repository could not be contacted. Make sure your server is allowed to contact www.dokuwiki.org and check your proxy settings. [Retry]"

### What I tried:
I have added the file httptest.php (linked in #1407) in the root directory of the wiki called it in my browser and received the following result:

```
connection pool 6.103515625E-5s
Array
(
)
opening connection 9.2029571533203E-5s
www.dokuwiki.org:80
request 0.026488065719604s
GET /lib/plugins/pluginrepo/api.php?q=dw2pdf&fmt=yaml HTTP/1.0
Accept-encoding: gzip
Accept: text/xml,application/xml,application/xhtml+xml,text/html,text/plain,image/png,image/jpeg,image/gif,*/*
Accept-Language: en-us
Host: www.dokuwiki.org
User-Agent: Mozilla/4.0 (compatible; DokuWiki HTTP Client; OpenBSD)
Connection: Keep-Alive

response headers 0.2035870552063s
HTTP/1.1 200 OK
Date: Fri, 05 Jun 2020 10:15:39 GMT
Server: Apache
Set-Cookie: DokuWiki=5m9mjmf6t1t0cueufd81een4pb; path=/; HttpOnly
Expires: Thu, 19 Nov 1981 08:52:00 GMT
Set-Cookie: DW68700bfd16c2027de7de74a5a8202a6f=deleted; expires=Thu, 01-Jan-1970 00:00:01 GMT; Max-Age=0; path=/; HttpOnly
Cache-Control: must-revalidate, post-check=0, pre-check=0
Pragma: public
X-Robots-Tag: noindex
Content-Type: text/yaml;charset=UTF-8
Content-Length: 53531
Connection: close

object(DokuHTTPClient)#18 (27) { ["agent"]=> string(55) "Mozilla/4.0 (compatible; DokuWiki HTTP Client; OpenBSD)" ["http"]=> string(3) "1.0" ["timeout"]=> int(15) ["cookies"]=> array(0) { } ["referer"]=> string(0) "" ["max_redirect"]=> int(3) ["max_bodysize"]=> int(0) ["max_bodysize_abort"]=> bool(true) ["header_regexp"]=> string(0) "" ["headers"]=> array(3) { ["Accept-encoding"]=> string(4) "gzip" ["Accept"]=> string(102) "text/xml,application/xml,application/xhtml+xml,text/html,text/plain,image/png,image/jpeg,image/gif,*/*" ["Accept-Language"]=> string(5) "en-us" } ["debug"]=> int(1) ["start"]=> float(1591352139.7317) ["keep_alive"]=> bool(true) ["error"]=> string(509) "Server returned bad answer HTTP/1.1 200 OK Date: Fri, 05 Jun 2020 10:15:39 GMT Server: Apache Set-Cookie: DokuWiki=5m9mjmf6t1t0cueufd81een4pb; path=/; HttpOnly Expires: Thu, 19 Nov 1981 08:52:00 GMT Set-Cookie: DW68700bfd16c2027de7de74a5a8202a6f=deleted; expires=Thu, 01-Jan-1970 00:00:01 GMT; Max-Age=0; path=/; HttpOnly Cache-Control: must-revalidate, post-check=0, pre-check=0 Pragma: public X-Robots-Tag: noindex Content-Type: text/yaml;charset=UTF-8 Content-Length: 53531 Connection: close " ["redirect_count"]=> int(0) ["status"]=> int(0) ["resp_body"]=> string(0) "" ["resp_headers"]=> array(0) { } ["user"]=> NULL ["pass"]=> NULL ["proxy_host"]=> string(0) "" ["proxy_port"]=> string(0) "" ["proxy_user"]=> string(0) "" ["proxy_pass"]=> string(0) "" ["proxy_ssl"]=> int(0) ["proxy_except"]=> string(0) "" ["boundary"]=> string(33) "---DokuWikiHTTPClient--4523452351" }
```

(I have also modified httptest.php to access the same URL with HTTPS instead of HTTP, with a very similar result.)

### Test conclusions
From the "HTTP/1.1 200 OK" result code and the content length I conclude, that the server is in fact able to access the plugin repository. However I do not know, why the error message is shown, and why no data from the repository is used for presenting the Plugins page.

### Open question:
Why is access to the plugin repository blocked and how can I fix this?

Thx.

**Comment by fiwswe** [comment]
OK, I have done another test on a local VM running a similar setup:

### Setup:
_(differences to the server with problem in **bold**)_
• OpenBSD 6.7 (same)
• Apache 2.4.43
• **PHP 7.3.18** (+ php-apache, php-gd Packages)
• Dokuwiki **2018-04-22b** "Greebo"
• **Server has a private IP in a LAN (with NAT to internet and no exposed host)**
• **No HTTP Basic authentication, access not restricted**

### Results:
This server does not have any problem accessing the plugin repository.

### Triage:
• OpenBSD is not the issue.
• Apache is not the issue.
• **PHP 7.4.6 might be the issue?**
• The minor version difference for Dokuwiki is most likely not the issue.
• I can't see there the access restriction might be an issue, as this test server is in a LAN and not reachable from the internet. Effectively it has even higher access restrictions.

### Further testing:
I have upgraded PHP to version 7.4.6 on my test VM and the problem now shows there as well. So this seems to be some **incompatibility with PHP 7.4.x**.

Is there a quick fix (before dokuwiki 2020-06-01 "Hogfather" is released and all of the plugins are updated for compatibility)?

**Comment by phy25** [comment]
Could you enable debug on the server that has the issue: https://www.dokuwiki.org/config:allowdebug

And open `/doku.php?id=start&do=admin&page=extension&tab=plugins&purge=1&httpdebug`?

You should see some debugging information on the top.

**Comment by fiwswe** [comment]
Sure:
set `$conf['allowdebug']  = 1;` in conf/local.php
Called the URL:
http://_SERVER_FQDN_/doku.php?id=start&do=admin&page=extension&tab=plugins&purge=1&httpdebug

This is the debug output:
```
connection pool 3.9100646972656E-5s
Array
(
)
opening connection 7.6055526733398E-5s
www.dokuwiki.org:80
request 0.026001930236816s
GET /lib/plugins/pluginrepo/api.php?cmd=ping HTTP/1.0
Accept-encoding: gzip
Accept: text/xml,application/xml,application/xhtml+xml,text/html,text/plain,image/png,image/jpeg,image/gif,*/*
Accept-Language: en-us
Host: www.dokuwiki.org
User-Agent: Mozilla/4.0 (compatible; DokuWiki HTTP Client; OpenBSD)
Connection: Keep-Alive

response headers 0.1241660118103s
HTTP/1.1 200 OK
Date: Sat, 06 Jun 2020 08:24:21 GMT
Server: Apache
Set-Cookie: DokuWiki=b49dthak5j96fnha8ou3q18d3e; path=/; HttpOnly
Expires: Thu, 19 Nov 1981 08:52:00 GMT
Set-Cookie: DW68700bfd16c2027de7de74a5a8202a6f=deleted; expires=Thu, 01-Jan-1970 00:00:01 GMT; Max-Age=0; path=/; HttpOnly
Cache-Control: must-revalidate, post-check=0, pre-check=0
Pragma: public
X-Robots-Tag: noindex
Content-Encoding: gzip
Content-Type: text/plain; charset=utf-8
Content-Length: 27
Connection: close
```

So the ping works ok. I don't see the debug info for the actual call to fetch the repository data though?


Hmm, next I looked at the code and for testing purposes I inserted 'return true;' as the first line of `function hasAccess()` in /lib/plugins/extension/helper/repository.php.

Result: Now it works! So something is causing a misinterpretation of the ping result it seems.

With HTTP debugging on I also see lots of debugging info showing access to the plugin repositoriy similar to this:
```
connection pool 6.6995620727539E-5s
Array
(
)
opening connection 9.9897384643555E-5s
www.dokuwiki.org:80
request 0.026806116104126s
GET /lib/plugins/pluginrepo/api.php?fmt=php&ext[]=acl HTTP/1.0
Accept-encoding: gzip
Accept: text/xml,application/xml,application/xhtml+xml,text/html,text/plain,image/png,image/jpeg,image/gif,*/*
Accept-Language: en-us
Host: www.dokuwiki.org
User-Agent: Mozilla/4.0 (compatible; DokuWiki HTTP Client; OpenBSD)
Connection: Keep-Alive

response headers 0.063391923904419s
HTTP/1.1 200 OK
Date: Sat, 06 Jun 2020 09:16:15 GMT
Server: Apache
Vary: Cookie
Set-Cookie: DokuWiki=0ps46gr90j4ep772mhsedvl8fi; path=/; HttpOnly
Expires: Thu, 19 Nov 1981 08:52:00 GMT
Set-Cookie: DW68700bfd16c2027de7de74a5a8202a6f=deleted; expires=Thu, 01-Jan-1970 00:00:01 GMT; Max-Age=0; path=/; HttpOnly
Cache-Control: must-revalidate, post-check=0, pre-check=0
Pragma: public
X-Robots-Tag: noindex
Connection: close
Content-Type: application/vnd.php.serialized

connection pool 4.5061111450195E-5s
Array
(
)
opening connection 8.0108642578125E-5s
www.dokuwiki.org:80
request 0.026947975158691s
GET /lib/plugins/pluginrepo/api.php?fmt=php&ext[]=authad HTTP/1.0
Accept-encoding: gzip
Accept: text/xml,application/xml,application/xhtml+xml,text/html,text/plain,image/png,image/jpeg,image/gif,*/*
Accept-Language: en-us
Host: www.dokuwiki.org
User-Agent: Mozilla/4.0 (compatible; DokuWiki HTTP Client; OpenBSD)
Connection: Keep-Alive

response headers 0.062964916229248s
HTTP/1.1 200 OK
Date: Sat, 06 Jun 2020 09:16:15 GMT
Server: Apache
Vary: Cookie
Set-Cookie: DokuWiki=v9ptt0ohv0o80iouipbtvc5akk; path=/; HttpOnly
Expires: Thu, 19 Nov 1981 08:52:00 GMT
Set-Cookie: DW68700bfd16c2027de7de74a5a8202a6f=deleted; expires=Thu, 01-Jan-1970 00:00:01 GMT; Max-Age=0; path=/; HttpOnly
Cache-Control: must-revalidate, post-check=0, pre-check=0
Pragma: public
X-Robots-Tag: noindex
Connection: close
Content-Type: application/vnd.php.serialized

…
```

**Comment by fiwswe** [comment]
I have traced the problem further to the function `sendRequest` in file inc/HTTPClient.php.

There in line 343 an exception is thrown 'Server returned bad answer HTTP/1.1 200 OK'. This is obviously not true so I suspect the preg_match call in the line above to be the culprit.

Indeed the following test code proves this theory:

File `test.php`:
```
<?php

define('kNL',
       "\n");
define('kCRLF',
       "\r\n");

echo 'PHP version: '.phpversion().kNL.kNL;

// Simulate the headers received:
$h = 'HTTP/1.1 200 OK'.kCRLF.
     'Date: Sat, 06 Jun 2020 10:12:20 GMT'.kCRLF.
     'Server: Apache'.kCRLF.
     'Vary: Cookie,Accept-Encoding'.kCRLF.
     'Set-Cookie: DokuWiki=b8obapn7ll4m88i98flj4dndqj; path=/; HttpOnly'.kCRLF.
     'Expires: Thu, 19 Nov 1981 08:52:00 GMT'.kCRLF.
     'Set-Cookie: DW68700bfd16c2027de7de74a5a8202a6f=deleted; expires=Thu, 01-Jan-1970 00:00:01 GMT; Max-Age=0; path=/; HttpOnly'.kCRLF.
     'Cache-Control: must-revalidate, post-check=0, pre-check=0'.kCRLF.
     'Pragma: public'.kCRLF.
     'X-Robots-Tag: noindex'.kCRLF.
     'Content-Encoding: gzip'.kCRLF.
     'Connection: close'.kCRLF.
     'Content-Type: text/plain; charset=utf-8'.kCRLF.
     kCRLF;

// Here is the questionable call:
$r = preg_match('/^HTTP\/(\d\.\d)\s*(\d+).*?\n/',$h,$m);

// Analyse the result:
var_dump($r);
var_dump($m);

if (!$r)
    echo 'Throw exception'.kNL;
else
    echo 'OK'.kNL;

?>
```

```
$ php test.php
PHP version: 7.3.18

int(1)
array(3) {
  [0]=>
  string(17) "HTTP/1.1 200 OK
"
  [1]=>
  string(3) "1.1"
  [2]=>
  string(3) "200"
}
OK
$ 
```

```
$ php-7.4 test.php
PHP version: 7.4.6

int(0)
array(0) {
}
Throw exception
$ 
```

Further experimenting showed that using the pattern:
`'/^HTTP\/(\d\.\d)\s*(\d+).*\r?\n/'` (note the **`\r`** in front of the `?`)
instead of the original pattern:
`'/^HTTP\/(\d\.\d)\s*(\d+).*?\n/'`
solves the problem. It works in both PHP 7.3.18 and PHP 7.4.6.

Actually I'm not sure what the `?` in the original pattern is supposed to accomplish anyway???

**Comment by fiwswe** [comment]
Please be aware that the current version of /inc/HTTP/HTTPClient.php in the GITHub repository still contains this faulty pattern. The same goes for the current release candidate.

After further testing someone might want to fix this…

**Comment by phy25** [comment]
Code mentioned:
https://github.com/splitbrain/dokuwiki/blob/af1080410dbed2076348f920c01902e6495dd405/inc/HTTP/HTTPClient.php#L332-L333

> Actually I'm not sure what the ? in the original pattern is supposed to accomplish anyway???

From https://regex101.com/:

- `.*?` matches any character (except for line terminators)
- `*?` Quantifier — Matches between zero and unlimited times, as few times as possible, expanding as needed (lazy)

I am interested in which exact PHP version breaks the previous regex or why it's not working (probably needs to run some Docker). Otherwise this seems to be a random fix to me.

**Comment by fiwswe** [comment]
Thanks for the regex explanation! I forgot about the greedy/lazy quantifier modifier.

https://regex101.com seems to be a good resource, although I see no way to simulate the CRLF line terminators from HTTP in the test string.

I think though, the PHP manual should be the authoritative source for regex patterns as implemented in PHP (unless the manual defers to an external source).

From the PHP manual: https://www.php.net/manual/en/regexp.reference.meta.php
".  match any character except newline (by default)"
Where _newline_ seems to be interpreted differently depending on PHP version.

From the PHP manual: https://www.php.net/manual/en/regexp.reference.onlyonce.php
`.*?`  or `*?` constructs are not mentioned.

From the PHP manual: https://www.php.net/manual/en/regexp.reference.repetition.php:
The `*?` construct is mentioned with respect to the _greediness_ of the quantifiers and the _question mark_ is as well: "However, if a quantifier is followed by a question mark, then it becomes lazy …". But also: "By default, the quantifiers are "greedy", …", which means the `?` turns the preceding `.*` into its lazy form, matching `.` the minimum number of times possible (as you mentioned).

That brings us down to the definition of `.` again. If `.` matches anything but `\n`(as the term _newline_ implies), then the construct `.*?` would indeed be correct. If however `\r` is also included in the definition of _newline_ then we would expect to see the results we see in PHP 7.4.6.


As to your question about the specific PHP version where the original pattern starts failing, OpenBSD only includes packages for the latest version of each, so I can't test e.g. PHP 7.4.0 - 7.4.5.


All of the above aside, the purpose of `.*?\n` in the pattern was to match anything between the HTTP status code (`200`) and the LF (`\n`) at the very end of the line including a possible CR (`\r`) right before the LF (to allow for CRLF as well as non-standard LF line endings).
While not strictly incorrect, this pattern would also match CRs not immediately preceding the LF, which would probably warrant the _"Server returned bad answer"_ exception to be thrown. So all in all I think my version of this part of the pattern (`.*\r?\n`) is actually more correct (assuming that `.` does not match CR, which again depends on the definition of _newline_). Maybe the pattern should be even more explicit: `[^\r]*\r?\n` to avoid the ambiguity of the definition of _newline_?
The complete pattern would then look like this: `/^HTTP\/(\d\.\d)\s*(\d+)[^\r]*\r?\n/`
I have tested this pattern successfully in PHP 7.3.18 and 7.4.6.

**Comment by fiwswe** [comment]
I found an older system with PHP 5.5.38 and tested the latest pattern successfully as well. (Actually the original, my first fix and the latest pattern all work correctly on this old version of PHP.) 

But testing with non-standard LF line terminators instead of CRLF the result was even better when using: `/^HTTP\/(\d\.\d)\s*(\d+)[^\r\n]*\r?\n/`
Note though that the result was not incorrect w.r.t. throwing the exception. Only the the first element of the matches array was different, including all of the header lines for the `[^\r]` case, and only the first header line for the `[^\r\n]` case. (An expected result for the greedy match of `[^\r]*`. when thinking about it)

Successfully tested with PHP 5.5.38, 7.3.18 and 7.4.6

**Comment by fiwswe** [comment]
Also tested successfully on PHP 5.6.30 (macOS 10.12.6) (both LF and CRLF versions, latest pattern with `[^\r\n]`)

**Comment by Klap-in** [comment]
If an unit test is built, than maybe Travis can be used as well for testing?  @phy, 7.4 is not yet part of the Travis list (e.g. https://travis-ci.org/github/splitbrain/dokuwiki/builds/695626002) Is that already possible?

**Comment by phy25** [comment]
@Klap-in We are blocked by #2927.

**Comment by phy25** [comment]
I tested based on your script (`/^HTTP\/(\d\.\d)\s*(\d+).*?\n/`) with the following Docker command:
```console
$ docker run --rm -it -v "$(pwd)":/usr/src/myapp -w /usr/src/myapp php:7.4.6 php ./test.php
```

I have 7.4.0, 7.4.5 and 7.4.6 tested and all of them are OK in Docker. I will then need to create a OpenBSD VM to see if it's an OpenBSD issue.

**Comment by fiwswe** [comment]
Interesting!

My OpenBSD 6.7 VM @home was relatively fresh, and I don't recall installing or modifying anything that might affect this. Still, getting an independent test done is probably a good idea.

In case this matters here is the list of manually installed packages from that VM:
```
20200607T234718 root@mfo1:/var/www/dokuwiki# pkg_info -am
arp-scan-1.9p13     ARP scanning and fingerprinting tool
colorls-6.5         ls(1) that can use color to display file attributes
dokuwiki-2018.04.22ap2 standards compliant, simple to use Wiki
emacs-26.3p1-no_x11 GNU editor: extensible, customizable, self-documenting
intel-firmware-20200508v0 microcode update binaries for Intel CPUs
lldpd-1.0.5         LLDP (802.1ab)/CDP/EDP/SONMP/FDP daemon and SNMP subagent
php-7.3.18          server-side HTML-embedded scripting language
php-7.4.6           server-side HTML-embedded scripting language
php-apache-7.3.18   php module for Apache httpd
php-apache-7.4.6    php module for Apache httpd
php-gd-7.3.18       image manipulation extensions for php
php-gd-7.4.6        image manipulation extensions for php
pkglocatedb-1.5     database of packages for use with locate(1)
postfix-3.6.20200312 fast, secure sendmail replacement
quirks-3.325        exceptions to pkg_add rules
20200607T234729 root@mfo1:/var/www/dokuwiki# 
```

The system is fully patched (-> syspatch(8)) and all packages are current (# pkg_add -u).

**Comment by phy25** [comment]
I got OpenBSD 6.7 and installed PHP by `pkg_add php-7.4.6` and can reproduce the error, which becomes interesting. Will try to see if I can install other minor versions to see what went wrong.

**Comment by phy25** [comment]
There is a 7.4.5 port (rather than 7.4.6) so I tried to build it following http://www.openbsd.org/faq/ports/ports.html, but it's taking too long (and my VM ran out of space twice) so I decided to give up. 

I am not familiar with PHP core or OpenBSD ports, so I don't think I am able to investigate further. However, if it's only not working on OpenBSD, I would say it's an environment issue and shouldn't be fixed simply by changing regular expression used.

**Comment by fiwswe** [comment]
OK, thanks for your efforts @phy25!

I have verified that the lowest version of PHP 7.4 that has a package for OpenBSD 6.7 is PHP 7.4.5 and a package file for amd64 (which my VM needs) can be found here:
https://ftp.openbsd.org/pub/OpenBSD/6.7/packages/amd64/php-7.4.5p0.tgz

I have never installed a package this way but possibly
`pkg_add https://ftp.openbsd.org/pub/OpenBSD/6.7/packages/amd64/php-7.4.5p0.tgz`
might work?

I tried this on my OpenBSD 6.7 VM @home (had to `pkg_delete php-7.4.6` and some packages dependent of it first) and then I tested with my test script. Same result for PHP 7.4.5, as for PHP 7.4.6:
- Original pattern with CRLF line endings: ERROR
- My last pattern with CRLF line endings: OK
- Original pattern with LF line endings: OK
- My last pattern with LF line endings: OK


Unfortunately no older versions of PHP 7.4 are packaged for OpenBSD 6.7, an no version of PHP 7.4 is packaged for OpenBSD 6.6 (or earlier). So the error must have been introduced with the php-7.4.5p0 package (either in the patches, or in the PHP base, or in some library in OpenBSD 6.7).

Further study of the documentation (https://man.openbsd.org/perlre and https://www.php.net/manual/en/reference.pcre.pattern.modifiers.php) let me try one more modified pattern:
`/^HTTP\/(\d\.\d)\s*(\d+).*?\n/s` (original pattern but note the added pattern modifier `s` for _PCRE_DOTALL_ at the end)
This works correctly with CRLF and LF line endings on PHP 7.4.5 (OpenBSD 6.7), 7.4.6 (OpenBSD 6.7), as well as 5.5.38 (OS X 10.11.6), 5.6.30 (macOS 10.12.6)

Even if this is a problem specific to PHP 7.4.x on OpenBSD 6.7 I'd vote for fixing it. Particularly considering the suggested fixes `/^HTTP\/(\d\.\d)\s*(\d+).*?\n/s` or `/^HTTP\/(\d\.\d)\s*(\d+)[^\r\n]*\r?\n/` are not incorrect in any way. But alas, it's not my code, so I'll live with having to patch this myself…

**Comment by phy25** [comment]
I guess OpenBSD PHP variant changed PCRE library or config somehow. Would be interesting if we could figure out what actually changed.

**Comment by fiwswe** [comment]
Well from `phpinfo();` on an OpenBSD 6.7, PHP 7.4.6 system I can see the following _(when searching for "pcre")_:

| Description | Value |
| --- | --- |
| **Configure Command:** |  … `'--with-pcre-jit=no'` … `'--with-external-pcre'` … |
| **PCRE (Perl Compatible Regular Expressions) Support** | enabled |
| **PCRE Library Version** | 10.34 2019-11-21 |
| **PCRE Unicode Version** | 12.1.0 |
| **PCRE JIT Support** | not compiled in |
| **pcre.backtrack_limit** | 1000000 |
| **pcre.recursion_limit** | 100000 |

**Comment by phy25** [comment]
```console
openbsd# php-7.4 -r "phpinfo();" | grep pcre -i
Configure Command =>  ... '--with-pcre-jit=no' ... '--with-external-pcre' ...
pcre
PCRE (Perl Compatible Regular Expressions) Support => enabled
PCRE Library Version => 10.34 2019-11-21
PCRE Unicode Version => 12.1.0
PCRE JIT Support => not compiled in
pcre.backtrack_limit => 1000000 => 1000000
pcre.recursion_limit => 100000 => 100000
openbsd# php-7.3 -r "phpinfo();" | grep pcre -i
Configure Command =>  ... '--with-pcre-jit=no' ... '--with-pcre-regex' ...
pcre
PCRE (Perl Compatible Regular Expressions) Support => enabled
PCRE Library Version => 10.32 2018-09-10
PCRE Unicode Version => 11.0.0
PCRE JIT Support => not compiled in
pcre.backtrack_limit => 1000000 => 1000000
pcre.recursion_limit => 100000 => 100000
```

```console
$ docker run --rm -it php:7.4.6 php -r "phpinfo();" | grep pcre -i
# Configure Command does not include anything regarding to PCRE
pcre
PCRE (Perl Compatible Regular Expressions) Support => enabled
PCRE Library Version => 10.34 2019-11-21
PCRE Unicode Version => 12.1.0
PCRE JIT Support => enabled
PCRE JIT Target => x86 64bit (little endian + unaligned)
pcre.backtrack_limit => 1000000 => 1000000
pcre.jit => 1 => 1
pcre.recursion_limit => 100000 => 100000
$ docker run --rm -it php:7.3 php -r "phpinfo();" | grep pcre -i
# Configure Command does not include anything regarding to PCRE
pcre
PCRE (Perl Compatible Regular Expressions) Support => enabled
PCRE Library Version => 10.32 2018-09-10
PCRE Unicode Version => 11.0.0
PCRE JIT Support => enabled
PCRE JIT Target => x86 64bit (little endian + unaligned)
pcre.backtrack_limit => 1000000 => 1000000
pcre.jit => 1 => 1
pcre.recursion_limit => 100000 => 100000
```

**Comment by phy25** [comment]
Reading https://www.pcre.org/original/doc/html/pcre_compile.html, I guess the native PCRE in OpenBSD does not enable `PCRE_DOTALL`, while PHP probably enables it by default. I would say @fiwswe's suggestion of adding `s` flag sounds good to me.

@fiwswe please feel free to submit a PR to get your credit! This should be easy through the GitHub web UI, though I am not sure if we have other regexp having this issue...

**Comment by fiwswe** [comment]
PR: #3173 on branch master created.

However I am new to GitHub. How do I create a PR for the release_stable_2018-04-22c tag? It seems PRs can only be done on branches, not tags?! And the code has been refactored in master compared to release_stable_2018-04-22c.

**Comment by phy25** [comment]
We will need to pick it up in a later release (not sure how we want to do it - a new release of Hogfather, or next release), but not in a currently tagged RC release.

I did some more research, even tracing down to [how OpenBSD built PCRE2](http://cvsweb.openbsd.org/cgi-bin/cvsweb/ports/devel/pcre2/Makefile.diff?r1=1.6&r2=1.11&f=h) in regards to `--enable-newline-*` (it always use `--enable-newline-is-any`),  with no luck. JIT does not seem to be a problem - if I disable JIT it still works fine under Ubuntu.

And while `PCRE_DOTALL` is closet to the current behavior of `/^HTTP\/(\d\.\d)\s*(\d+).*?\n/` in other environment, you are probably right before that it might be better to change how we match this. I thought about `/^HTTP\/(\d\.\d)\s*(\d+).*?[\r\n]/`...

**Comment by fiwswe** [comment]
Update: OpenBSD 6.7 packages PHP 7.3.19 & 7.4.7 have been released.

PHP 7.4.7 behaves the same as 7.4.6 on OpenBSD 6.7. The proposed fix #3173 still works.
PHP 7.3.19 is still unaffected either way.

I had not noticed this before but for PHP 7.3.19 on OpenBSD 6.7 the configure command contains: `… '--with-pcre-jit=no' … '--with-pcre-regex' …` which is different from `… '--with-pcre-jit=no' … '--with-external-pcre' …` for version 7.4.6-7.4.7.

According to https://www.php.net/manual/en/pcre.installation.php `--with-pcre-regex` also specifies an external PCRE library. https://bugs.php.net/bug.php?id=79479 mentions the changed option (and the missing documentation). Effectively it does not make much of a difference that I can see.

The PCRE versions are identical to those in your docker tests -> https://github.com/splitbrain/dokuwiki/issues/3148#issuecomment-643696826 on OpenBSD 6.7.

**Comment by Antiokh** [comment]
I have the same issue on the server, but no "response" section in debug mode. How can I fix it?
